New post: SafeCoder vs. Closed-source Services (#1447)
huggingface/blog: safecoder-vs-closed-models.md

---

> title: "XXX"

Reviewer: ```suggestion
title: "SafeCoder vs. Closed-source Code Assistants"
```

We also need to add the entry in the `_blog.yml` file.

> thumbnail: /blog/assets/safecoder-vs-closed-source-services/XXX.png

Reviewer: Just a reminder to update when we have it :)

authors:
- user: juliensimon
---

> # SafeCoder vs. Closed-source Services

Reviewer 2: ```suggestion
# SafeCoder vs. Closed-source Code Assistants
```

Reviewer 2: for people who don't know SafeCoder is a code assistant?

<!-- {blog_metadata} -->
<!-- {authors} -->

> For decades, software developers have designed methodologies, processes, and tools that help them improve code quality and increase productivity. For instance, Agile, test-driven development, code reviews, CI/CD, and more are now staples in the software industry. 

Reviewer 3: ```suggestion
For decades, software developers have designed methodologies, processes, and tools that help them improve code quality and increase productivity. For instance, agile, test-driven development, code reviews, CI/CD, and more are now staples in the software industry. 
```

In "How Google Tests Software" (Addison-Wesley, 2012), Google reports that fixing a bug during system tests, the final testing stage, is 1000x more expensive than fixing it at the unit testing stage. This puts a lot of pressure on developers, the first link in the chain, to write quality code from the get-go. 

> For all the hype surrounding generative AI, code generation seems a promising way to help developers deliver better code. Indeed, early studies show that managed services like [GitHub Copilot](https://github.blog/2023-06-27-the-economic-impact-of-the-ai-powered-developer-lifecycle-and-lessons-from-github-copilot) or [Amazon CodeWhisperer](https://aws.amazon.com/codewhisperer/) help developers be more productive.

Reviewer 3: ```suggestion
For all the hype surrounding generative AI, code generation seems a promising way to help developers deliver better code fast. Indeed, early studies show that managed services like [GitHub Copilot](https://github.blog/2023-06-27-the-economic-impact-of-the-ai-powered-developer-lifecycle-and-lessons-from-github-copilot) or [Amazon CodeWhisperer](https://aws.amazon.com/codewhisperer/) help developers be more productive.
```

However, these services rely on closed-source models that can't be customized to your technical culture and processes. Hugging Face released [SafeCoder](https://huggingface.co/blog/starcoder) a few weeks ago to fix this. SafeCoder is a code assistant solution built for the enterprise that gives you state-of-the-art models, transparency, customizability, IT flexibility, and privacy.

In this post, we'll compare SafeCoder to closed-source services and highlight the benefits you can expect from our solution.

## State-of-the-art models

> SafeCoder relies on the [StarCoder](https://huggingface.co/blog/starcoder) models, a family of open-source models designed and trained within the [BigCode](https://huggingface.co/bigcode) collaborative project.

Reviewer 3: ```suggestion
SafeCoder is currently built om top of the [StarCoder](https://huggingface.co/blog/starcoder) models, a family of open-source models designed and trained within the [BigCode](https://huggingface.co/bigcode) collaborative project.
```

> StarCoder is a 15.5 billion parameter model trained for code generation in over 80 programming languages. It uses a state-of-the-art architecture called Multi-Query Attention (MQA) to improve throughput and reduce latency, a technique also present in the [Falcon](https://huggingface.co/blog/falcon) and [LLaMa 2](https://huggingface.co/blog/llama2) models.

Reviewer 3: ```suggestion
StarCoder is a 15.5 billion parameter model trained for code generation in over 80 programming languages. It uses innovative architectural concepts, like Multi-Query Attention (MQA), to improve throughput and reduce latency, a technique also present in the [Falcon](https://huggingface.co/blog/falcon) and adapted for [LLaMa 2](https://huggingface.co/blog/llama2) models.
```

llama uses GQA which a bit different than MQA

> StarCoder has an 8-kilobyte context window, helping it take into account more of your existing code to generate new code. It can also do fill-in-the-middle, i.e., insert within existing code, instead of just appending new code at the end.

Reviewer 3: ```suggestion
StarCoder has an 8192 token context window, helping it take into account more of your existing code to generate new code. It can also do fill-in-the-middle, i.e., insert within existing code, instead of just appending new code at the end.
```

Author: good catch!

Lastly, just like [HuggingChat](https://huggingface.co/chat/), SafeCoder will introduce new state-of-the-art models over time, giving you a seamless upgrade path.

> Unfortunately, closed-source services don't share information about the underlying models and their capabilities. 

Reviewer 2: ```suggestion
Unfortunately, closed-source code assistant services don't share information about the underlying models, their capabilities and their training data. 
```

## Transparency

In line with the [Chinchilla Scaling Law](https://arxiv.org/abs/2203.15556v1), SafeCoder is a compute-optimal model trained on 1 trillion (1,000 billion) code tokens. These tokens are extracted from [The Stack](https://huggingface.co/datasets/bigcode/the-stack), a 2.7 terabyte dataset built from permissively licensed open-source repositories. 
All efforts are made to honor opt-out requests, and we built a [tool](https://huggingface.co/spaces/bigcode/in-the-stack) that lets repository owners check if their code is part of the dataset.

> As expected, our [research paper](https://arxiv.org/abs/2305.06161) discloses the model architecture, the training process, and detailed metrics.

Reviewer: ```suggestion
Our [research paper](https://arxiv.org/abs/2305.06161) fully discloses the model architecture, the training process, and provides detailed metrics.
```

Unfortunately, closed-source services are only available as managed services.

## Security and privacy

Security is always a top concern, all the more when source code is involved. Intellectual property and privacy must be protected at all costs.

Whether you run on-premise or in the cloud, SafeCoder is 100% under your administrative control. You can apply and monitor your security controls and maintain strong and consistent compliance across your IT platform.

SafeCoder doesn't spy on any of your data. Your prompts and suggestions are yours and yours only. SafeCoder doesn't call home and send telemetry data to Hugging Face or anyone else. No one but you needs to know how and when you're using SafeCoder. In fact, SafeCoder doesn't even require an Internet connection. You can (and should) run it fully air-gapped.

> Closed-source services rely on the security of the underlying cloud. Whether this works or not for your compliance posture is your call. For enterprise users, prompts and suggestions are not stored (they are for individual users). However, we regret to point out that GitHub collects  ["user engagement data](https://docs.github.com/en/copilot/overview-of-github-copilot/about-github-copilot-for-business)" with no possibility to opt-out. AWS does the same by default but lets you [opt out](https://docs.aws.amazon.com/codewhisperer/latest/userguide/sharing-data.html).

Reviewer: ```suggestion
Closed-source services rely on the security of the underlying cloud. Whether this works or not for your compliance posture is your call. For enterprise users, prompts and suggestions are not stored (they are for individual users). However, we regret to point out that GitHub collects  ["user engagement data"](https://docs.github.com/en/copilot/overview-of-github-copilot/about-github-copilot-for-business) with no possibility to opt-out. AWS does the same by default but lets you [opt out](https://docs.aws.amazon.com/codewhisperer/latest/userguide/sharing-data.html).
```
